lib/token-store.js

```diff
@@ -5,199 +5,1071 @@
+            map[range[0]] = it;
+            map[range[1] - 1] = it;
+            it += 1;
+        }
+
+        nextStart = (it < tokens.length) ? tokens[it].range[0] : Number.MAX_SAFE_INTEGER;
+        while (ic < comments.length && (range = comments[ic].range)[0] < nextStart) {
+            map[range[0]] = it;
+            map[range[1] - 1] = it;
+            ic += 1;
+        }
+    }
+
+    return map;
+}
+
+/**
+ * Creates the cursor iterates tokens with normalized options.
+ *
+ * @param {CursorPair} cursors - The cursor pair to initialize base cursor.
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @param {Object} indexMap - The map from locations to indices in `tokens`.
+ * @param {number} startLoc - The start location of the iteration range.
+ * @param {number} endLoc - The end location of the iteration range.
+ * @param {boolean} includeComments - The flag to iterate comments as well.
+ * @param {Function|null} filter - The predicate function to choose tokens.
+ * @param {number} skip - The count of tokens the cursor skips.
+ * @param {number} count - The maximum count of tokens the cursor iterates. Zero is no iteration for backward compatibility.
+ * @returns {Cursor} The created cursor.
+ * @private
+ */
+function createCursor(cursors, tokens, comments, indexMap, startLoc, endLoc, includeComments, filter, skip, count) {
+    let cursor = cursors.createCursor(tokens, comments, indexMap, startLoc, endLoc, includeComments);
+
+    if (filter) {
+        cursor = new FilterCursor(cursor, filter);
+    }
+    if (skip >= 1) {
+        cursor = new SkipCursor(cursor, skip);
+    }
+    if (count >= 0) {
+        cursor = new LimitCursor(cursor, count);
+    }
+
+    return cursor;
+}
+
+/**
+ * Creates the cursor iterates tokens with options.
+ *
+ * @param {CursorPair} cursors - The cursor pair to initialize base cursor.
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @param {Object} indexMap - The map from locations to indices in `tokens`.
+ * @param {number} startLoc - The start location of the iteration range.
+ * @param {number} endLoc - The end location of the iteration range.
+ * @param {number|Function|Object} [opts=0] - The option object. If this is a number then it's `opts.skip`. If this is a function then it's `opts.filter`.
+ * @param {boolean} [opts.includeComments=false] - The flag to iterate comments as well.
+ * @param {Function|null} [opts.filter=null] - The predicate function to choose tokens.
+ * @param {number} [opts.skip=0] - The count of tokens the cursor skips.
+ * @returns {Cursor} The created cursor.
+ * @private
+ */
+function createCursorWithSkip(cursors, tokens, comments, indexMap, startLoc, endLoc, opts) {
+    let includeComments = false;
+    let skip = 0;
+    let filter = null;
+
+    if (typeof opts === "number") {
+        skip = opts | 0;
```

Comment:
I've actually never personally used the bitwise OR operator. Just for my own understanding, is this being used for performance reasons (as opposed to `Math.floor()`, for instance)?